src/Makefile.am

```diff
@@ -52,8 +54,9 @@ if ENABLE_NEURON
 else
   lib_LTLIBRARIES = libnccl-net.la
   libnccl_net_la_SOURCES =
-  libnccl_net_la_LIBADD = libinternal_net_plugin.la
-  libnccl_net_la_LDFLAGS = -module -avoid-version
+  libnccl_net_la_LIBADD = libinternal_net_plugin.la $(CUDA_LIBS)
+  libnccl_net_la_LDFLAGS = -module -avoid-version $(CUDA_LDFLAGS)
+  libnccl_net_la_CPPFLAGS = $(CUDA_CPPFLAGS)
```

Comment:
why are you adding CUDA_CPPFLAGS to both the whole-makefile flags and the per-library CPPFLAGS?  This is confusing overkill.

Reply:
If I remove them from the whole-makefile flags I get:
                           
```                        
make[2]: Entering directory 'nccl-net-ofi/tests/functional'
depbase=`echo nccl_connection.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
mpicc -DHAVE_CONFIG_H -I. -I../../include  -I../../include     -O3 -Wall -Wc++-compat -MT nccl_connection.o -MD -MP -MF $depbase.Tpo -c -o nccl_connection.o nccl_connection.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo nccl_message_transfer.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
mpicc -DHAVE_CONFIG_H -I. -I../../include  -I../../include     -O3 -Wall -Wc++-compat -MT nccl_message_transfer.o -MD -MP -MF $depbase.Tpo -c -o nccl_message_transfer.o nccl_message_transfer.c &&\
mv -f $depbase.Tpo $depbase.Po
depbase=`echo ring.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
mpicc -DHAVE_CONFIG_H -I. -I../../include  -I../../include     -O3 -Wall -Wc++-compat -MT ring.o -MD -MP -MF $depbase.Tpo -c -o ring.o ring.c &&\
mv -f $depbase.Tpo $depbase.Po
bash ../../libtool  --tag=CC   --mode=link mpicc  -O3 -Wall -Wc++-compat    -o nccl_connection nccl_connection.o ../../src/libinternal_net_plugin.la   -lhwloc -llttng-ust -lfabric -lm     
libtool: link: mpicc -O3 -Wall -Wc++-compat -o nccl_connection nccl_connection.o  ../../src/.libs/libinternal_net_plugin.a -lhwloc -llttng-ust -lfabric -lm
/bin/ld: ../../src/.libs/libinternal_net_plugin.a(nccl_ofi_cuda.o): undefined reference to symbol 'cudaGetDriverEntryPoint@@libcudart.so.12'
/bin/ld: /lib/libcudart.so.12: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:438: nccl_connection] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory 'nslick/nccl-net-ofi/tests/functional'
make[1]: *** [Makefile:377: all-recursive] Error 1
make[1]: Leaving directory 'nslick/nccl-net-ofi/tests'
make: *** [Makefile:430: all-recursive] Error 1
```
 
If I instead remove them from libnccl_net_la I get:
                                   
```                                
bash ../../libtool  --tag=CC   --mode=link mpicc  -O3 -Wall -Wc++-compat    -o ring ring.o ../../src/libinternal_net_plugin.la   -lhwloc -llttng-ust -lfabric -lm 
ld: ../../src/.libs/libinternal_net_plugin.a(nccl_ofi_cuda.o): undefined reference to symbol 'cudaGetDriverEntryPoint@@libcudart.so.12'
ld: lib/libcudart.so.12: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:438: nccl_connection] Error 1
make[2]: *** Waiting for unfinished jobs....
libtool: link: mpicc -O3 -Wall -Wc++-compat -o ring ring.o  ../../src/.libs/libinternal_net_plugin.a -lhwloc -llttng-ust -lfabric -lm
ld: ../../src/.libs/libinternal_net_plugin.a(nccl_ofi_cuda.o): undefined reference to symbol 'cudaGetDriverEntryPoint@@libcudart.so.12'
lib/libcudart.so.12: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:446: ring] Error 1
make[2]: Leaving directory 'nccl-net-ofi/tests/functional'
make[1]: *** [Makefile:377: all-recursive] Error 1
make[1]: Leaving directory 'nccl-net-ofi/tests'
make: *** [Makefile:430: all-recursive] Error 1
```

Reply:
The top of your branch plus the following commit works for me

```
% git diff                                                                    ✹
diff --git a/m4/check_pkg_cuda.m4 b/m4/check_pkg_cuda.m4
index f9f2d2b..674d5da 100644
--- a/m4/check_pkg_cuda.m4
+++ b/m4/check_pkg_cuda.m4
@@ -24,7 +24,7 @@ AC_DEFUN([CHECK_PKG_CUDA], [
         [AS_IF([test -d $(realpath ${with_cuda})/lib64], [check_pkg_libdir="lib64"], [check_pkg_libdir="lib"])
          CUDA_LDFLAGS="-L$(realpath ${with_cuda})/${check_pkg_libdir}"
          CUDA_LIBS="-lcudart_static -lrt -ldl"
-         CUDA_CPPFLAGS="-fPIC -I$(realpath $(realpath ${with_cuda})/include)"
+         CUDA_CPPFLAGS="-I$(realpath $(realpath ${with_cuda})/include)"
          LDFLAGS="${CUDA_LDFLAGS} ${LDFLAGS}"
          LIBS="${CUDA_LIBS} ${LIBS}"
          CPPFLAGS="${CUDA_CPPFLAGS} ${CPPFLAGS}"
diff --git a/src/Makefile.am b/src/Makefile.am
index c2ce05b..2353d80 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -4,7 +4,7 @@
 # See LICENSE.txt for license information
 #
 
-AM_CPPFLAGS = -I$(top_srcdir)/include -DXML_DIR=\"${pkgdatadir}/xml\"
+AM_CPPFLAGS = -I$(top_srcdir)/include -DXML_DIR=\"${pkgdatadir}/xml\" $(CUDA_CPPFLAGS)
 
 #
 # net plugin
@@ -34,8 +34,6 @@ endif
 if ENABLE_NEURON
   sources += nccl_ofi_interface_neuron.c
 else
-  AM_CPPFLAGS += $(CUDA_CPPFLAGS)
-  AM_LDFLAGS = $(CUDA_LDFLAGS)
   sources += nccl_ofi_cuda.c \
        nccl_ofi_interface_nvidia.c
 endif
@@ -57,7 +55,6 @@ else
   libnccl_net_la_SOURCES =
   libnccl_net_la_LIBADD = libinternal_net_plugin.la $(CUDA_LIBS)
   libnccl_net_la_LDFLAGS = -module -avoid-version $(CUDA_LDFLAGS)
-  libnccl_net_la_CPPFLAGS = $(CUDA_CPPFLAGS)
 endif
 
 
```